src/RollerPeriphery.sol

```diff
@@ -4,93 +4,159 @@ pragma solidity 0.8.13;
     /// @param receiver Destination address for the returned assets
     /// @param maxSharesOut Maximum amount of shares burned
     /// @return shares Number of shares to redeem
-    function withdraw(ERC4626 vault, uint256 assets, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
-        if ((shares = vault.withdraw(assets, receiver, msg.sender)) > maxSharesOut) {
+    function withdraw(AutoRoller roller, uint256 assets, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
+        if ((shares = roller.withdraw(assets, receiver, msg.sender)) > maxSharesOut) {
+            revert MaxSharesError();
+        }
+    }
+
+    function withdrawUnderlying(AutoRoller roller, uint256 underlyingOut, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
+        AdapterLike adapter = AdapterLike(address(roller.adapter()));
+
+        uint256 assetOut = underlyingOut.divWadDown(adapter.scale()); // Tokens out, round down.
+
+        if ((shares = roller.withdraw(assetOut, address(this), msg.sender)) > maxSharesOut) {
             revert MaxSharesError();
         }
+
+        ERC20(adapter.underlying()).safeTransfer(
+            receiver,
+            adapter.unwrapTarget(roller.asset().balanceOf(address(this)))
+        );
     }
 
     /// @notice Mint vault shares with slippage protection 
-    /// @param vault ERC4626 vault
+    /// @param roller AutoRoller vault
     /// @param shares Number of shares to mint
     /// @param receiver Destination address for the returned shares
     /// @param maxAmountIn Maximum amount of assets pulled from msg.sender
     /// @return assets Amount of asset pulled from msg.sender and used to mint vault shares
-    function mint(ERC4626 vault, uint256 shares, address receiver, uint256 maxAmountIn) external returns (uint256 assets) {
-        ERC20(vault.asset()).safeTransferFrom(msg.sender, address(this), vault.previewMint(shares));
+    function mint(AutoRoller roller, uint256 shares, address receiver, uint256 maxAmountIn) external returns (uint256 assets) {
+        ERC20(roller.asset()).safeTransferFrom(msg.sender, address(this), roller.previewMint(shares));
 
-        if ((assets = vault.mint(shares, receiver)) > maxAmountIn) {
+        if ((assets = roller.mint(shares, receiver)) > maxAmountIn) {
             revert MaxAssetError();
         }
     }
 
+    function mintFromUnderlying(AutoRoller roller, uint256 shares, address receiver, uint256 maxAmountIn) external returns (uint256 underlyingIn) {
+        AdapterLike adapter = AdapterLike(address(roller.adapter()));
+
+        ERC20(adapter.underlying()).safeTransferFrom(
+            msg.sender,
+            address(this),
+            underlyingIn = roller.previewMint(shares).mulWadUp(adapter.scale()) // Tokens in, round up.
+        );
+        adapter.wrapUnderlying(underlyingIn);
+
+        uint256 targetIn = roller.mint(shares, receiver);
```

Comment:
Recommend rebranding to `assetIn` for consistency with other methods

Reply:
done!